torchmdnet/module.py

```diff
@@ -57,34 +57,95 @@ def configure_optimizers(self):
 
     def validation_step(self, batch, batch_idx, *args):
-        if len(args) == 0 or (len(args) > 0 and args[0] == 0):
-            # validation step
-            return self.step(batch, mse_loss, "val")
-        # test step
-        return self.step(batch, l1_loss, "test")
+        # If args is not empty the first (and only) element is the dataloader_idx
+        # We want to test every number of epochs just for reporting, but this is not supported by Lightning.
+        # Instead, we trick it by providing two validation dataloaders and interpreting the second one as test.
+        # The dataloader takes care of sending the two sets only when the second one is needed.
+        is_val = len(args) == 0 or (len(args) > 0 and args[0] == 0)
+        step_type = (
+            {"loss_fn_list": [l1_loss, mse_loss], "stage": "val"}
+            if is_val
+            else {"loss_fn_list": [l1_loss], "stage": "test"}
+        )
+        return self.step(batch, **step_type)
 
     def test_step(self, batch, batch_idx):
-        return self.step(batch, l1_loss, "test")
+        return self.step(batch, [l1_loss], "test")
+
+    def _compute_losses(self, y, neg_y, batch, loss_fn, stage):
+        # Compute the loss for the predicted value and the negative derivative (if available)
+        # Args:
+        #   y: predicted value
+        #   neg_y: predicted negative derivative
+        #   batch: batch of data
+        #   loss_fn: loss function to compute
+        # Returns:
+        #   loss_y: loss for the predicted value
+        #   loss_neg_y: loss for the predicted negative derivative
+        loss_y, loss_neg_y = 0.0, 0.0
+        loss_name = loss_fn.__name__
+        if self.hparams.derivative and "neg_dy" in batch:
+            loss_neg_y = loss_fn(neg_y, batch.neg_dy)
+            loss_neg_y = self._update_loss_with_ema(
+                stage, "neg_dy", loss_name, loss_neg_y
+            )
+        if "y" in batch:
+            loss_y = loss_fn(y, batch.y)
+            loss_y = self._update_loss_with_ema(stage, "y", loss_name, loss_y)
+        return {"y": loss_y, "neg_dy": loss_neg_y}
+
+    def _update_loss_with_ema(self, stage, type, loss_name, loss):
+        # Update the loss using an exponential moving average when applicable
+        # Args:
+        #   stage: stage of the training (train, val, test)
+        #   type: type of loss (y, neg_dy)
+        #   loss_name: name of the loss function
+        #   loss: loss value
+        alpha = getattr(self.hparams, f"ema_alpha_{type}")
+        if stage in ["train", "val"] and alpha < 1:
+            ema = (
+                self.ema[stage][type][loss_name]
+                if loss_name in self.ema[stage][type]
+                else loss.detach()
+            )
+            loss = alpha * loss + (1 - alpha) * ema
+            self.ema[stage][type][loss_name] = loss.detach()
+        return loss
 
-    def step(self, batch, loss_fn, stage):
+    def _append_to_dict(self, d, k, v):
```

Comment:
check out https://docs.python.org/3/library/collections.html#collections.defaultdict

Reply:
Cool! did not know about this one